server/conf/evolutions/default/44.sql

```diff
@@ -0,0 +1,66 @@
+# --- Add database triggers to set the latest_status column on creating an application_event or
+# --- updating an application.
+# --- For rationale behind choosing database triggers, see:
+# ---   https://github.com/civiform/civiform/issues/3269#issuecomment-1232193163
+# --- Relevant documentation:
+# --- Trigger functions: https://www.postgresql.org/docs/current/sql-createtrigger.html
+# --- Create trigger statement: https://www.postgresql.org/docs/current/sql-createtrigger.html
+
+# --- !Ups
+
+alter table applications add column if not exists latest_status varchar;
+
+-- The latest status associated with an application is set to NULL if no application event entries
+-- exist. Otherwise, the text from the latest event corresponding to a status change is used.
+CREATE OR REPLACE FUNCTION retrieve_latest_status(app_id applications.id%TYPE) RETURNS text AS $$
+  SELECT
+      (CASE WHEN details->'status_event'->>'status_text' = ''
+        THEN NULL
+        ELSE details->'status_event'->>'status_text'
+       END) AS latest_status
+      FROM application_events
+      WHERE
+        application_id = app_id AND
+        details->>'event_type' = 'STATUS_CHANGE'
+      ORDER BY create_time DESC
+      LIMIT 1
+$$ LANGUAGE SQL;
+
+-- A function that sets the "latest_status" column anytime an application is updated.
+-- Note: The "NEW" variable corresponds to the application record to be inserted / updated.
+CREATE OR REPLACE FUNCTION process_application_change() RETURNS TRIGGER AS $$
+  BEGIN
+    NEW.latest_status := retrieve_latest_status(NEW.id);;
+    RETURN NEW;;
+  END;;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER application_change BEFORE INSERT OR UPDATE ON applications
```

Comment:
I'm not sure why we need to trigger this on insert or update.

For insert it should be null which we can make the default in the add column statement.

For Update, won't we copy forward the existing value with each update?  Or Is it that every update overwrites this to null and then the trigger needs to fix it?

Eitherway can you doc the data-flow/rationale clearer?

Reply:
Moved both trigger definitions down side-by-side and attempted to document. As currently implemented, all is as you said. However, this trigger is specifically here to guard against application code attempting to change that behavior by explicitly modifying the latest_status column. The trigger guards against that being supported (and has unit tests to assert that). We can drop this in the future if we have a real need to modify the column in application code. However, that'll open us up to all of the concerns around ensuring that the column is manually kept in sync within application code.